apps/wallet-mobile/src/features/WalletManager/context/state.ts

```diff
@@ -0,0 +1,100 @@
+import {Chain, Wallet} from '@yoroi/types'
+import {NetworkManager} from '@yoroi/types/lib/typescript/network/manager'
+import {castDraft, freeze, produce} from 'immer'
+
+import {throwLoggedError} from '../../../kernel/logger/helpers/throw-logged-error'
+import {logger} from '../../../kernel/logger/logger'
+import {YoroiWallet} from '../../../yoroi-wallets/cardano/types'
+import {networkManagers} from '../common/constants'
+import {WalletManager} from '../wallet-manager'
+
+export const walletManagerReducer = (state: WalletManagerState, action: WalletManagerAction) => {
+  return produce(state, (draft) => {
+    switch (action.type) {
+      case WalletManagerActionType.NetworkSelected:
+        draft.selected.network = action.network
+        draft.selected.networkManager = castDraft(networkManagers[action.network])
+        break
+
+      case WalletManagerActionType.WalletSelected:
+        draft.selected.wallet = castDraft(action.wallet)
+        draft.selected.meta = action.meta
+        break
+
+      case WalletManagerActionType.SelectedMetaUpdated:
+        // only cares if selected one has been updated
+        if (draft.selected.meta != null) {
+          const newMeta = action.metas.get(draft.selected.meta.id)
+          if (newMeta != null) {
+            draft.selected.meta = newMeta
+          } else {
+            logger.error('walletManagerReducer: selected meta is gone')
+          }
+        }
+        break
+    }
+  })
+}
+
+export type WalletManagerAction =
+  | {
+      type: WalletManagerActionType.NetworkSelected
+      network: Chain.SupportedNetworks
+    }
+  | {
+      type: WalletManagerActionType.WalletSelected
+      wallet: YoroiWallet | null
+      meta: Wallet.Meta | null
+    }
+  | {
+      type: WalletManagerActionType.SelectedMetaUpdated
+      metas: Map<YoroiWallet['id'], Wallet.Meta>
+    }
+
+export type WalletManagerState = {
+  selected: {
+    wallet: YoroiWallet | null
+    meta: Wallet.Meta | null
+    network: Chain.SupportedNetworks
```

Comment:
isMainnet is a property of networkManager, network is just an id (string) instead of number that was misleading to chainId, there are 2 ways to access it, 1 by the selected networkManager, another with the selected wallet, the better way is always through the wallet cuz the selected wallet can be switch while some async operation is running in the current one. So the old API is still there -> wallet.isMainnet, which translate to the networkManager injected in the while -> wallet.networkManager.isMainnet. (getter)